core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -63,6 +63,7 @@
   private final CallOptions callOptions;
   private ClientStream stream;
   private volatile ScheduledFuture<?> deadlineCancellationFuture;
+  private volatile boolean inOnReady;
```

Comment:
This seems racy. If the application is writing on another thread (let's say it is using the async stub, and is ignoring the onReady() callbacks), then that other thread could see `inOnReady == true` after `stream.flush()` has been called, leading to data being stuck in the buffer. Note that this is not solved by setting `inOnReady = false` before calling `stream.flush()`.